File: 3186-maximum-total-damage-with-spell-casting/3186-maximum-total-damage-with-spell-casting.cpp
class Solution {
public:
    long long maximumTotalDamage(vector<int>& power) {
        map<int, int> count;

        for (int elem : power) ++count[elem];
        
        vector<pair<int, int>> spells{{-1e9, 0}};

        for (auto& p : count) spells.push_back(p);

        int n = spells.size();
        long long maxDamage = 0;
        vector<long long> f(n, 0);
        for (int i = 1, j = 1; i < n; ++i){

            while (j < i && spells[j].first < spells[i].first - 2)
                maxDamage = max(maxDamage, f[j++]);

            f[i] = maxDamage + 1LL * spells[i].first * spells[i].second;
        }

        return *max_element(f.begin(), f.end());
    }

};

File: 3186-maximum-total-damage-with-spell-casting/3186-maximum-total-damage-with-spell-casting.py
class Solution:
    def maximumTotalDamage(self, power: List[int]) -> int:
        count = Counter(power)
        spells = [(-(10 ** 9), 0)]

        for k in sorted(count.keys()):
            spells.append((k, count[k]))
        
        n = len(spells)
        f, mx, j = [0] * n, 0, 1

        for i in range(1, n):
            while j < i and spells[j][0] < spells[i][0] - 2:
                mx = max(mx, f[j])
                j += 1
            f[i] = mx + spells[i][0] * spells[i][1]
        
        return max(f)